sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -359,7 +359,14 @@ object QueryExecution {
   private[execution] def prepareForExecution(
       preparations: Seq[Rule[SparkPlan]],
       plan: SparkPlan): SparkPlan = {
-    preparations.foldLeft(plan) { case (sp, rule) => rule.apply(sp) }
+    val planChangeLogger = new PlanChangeLogger[SparkPlan]()
```

Comment:
We also run physical rules in `AdaptiveSparkPlanExec`, shall we log plan change there as well?

Reply:
Yea, sure. I'll make a PR later.

Reply:
https://github.com/apache/spark/pull/29774